File: src/main/java/com/github/bjoernpetersen/jmusicbot/provider/Suggester.java
package com.github.bjoernpetersen.jmusicbot.provider;

import com.github.bjoernpetersen.jmusicbot.IdPlugin;
import com.github.bjoernpetersen.jmusicbot.InitStateWriter;
import com.github.bjoernpetersen.jmusicbot.InitializationException;
import com.github.bjoernpetersen.jmusicbot.Song;
import com.github.bjoernpetersen.jmusicbot.playback.SongEntry;
import java.util.List;
import javax.annotation.Nonnull;

public interface Suggester extends IdPlugin, Dependent<Provider> {

  /**
   * <p>Suggest the next song to play.</p>
   *
   * <p>When a song has been returned by this method, it can be assumed it's going to be played.
   * This means it should be removed from the 'next suggestions' list.</p>
   *
   * @return a song to play
   * @throws BrokenSuggesterException if the suggester can't suggest anything
   */
  @Nonnull
  Song suggestNext() throws BrokenSuggesterException;

  /**
   * <p>Gets a list of next suggestions which will be returned by calling {@link
   * #suggestNext()}.</p>
   *
   * <p>The returned list must contain at least one element.</p>
   *
   * @param maxLength the maximum length of the returned list
   * @return a list of next suggestions
   * @throws BrokenSuggesterException if the suggester can't suggest anything
   */
  @Nonnull
  List<Song> getNextSuggestions(int maxLength) throws BrokenSuggesterException;

  /**
   * <p>Notifies this Suggester that the specified song entry has been played.</p>
   *
   * <p>It is recommended that the song will be removed from the next suggestions when this method
   * is called.</p>
   *
   * <p>It is guaranteed that the song comes from a provider this suggester depends on.</p>
   *
   * <p>The default implementation calls {@link #removeSuggestion(Song)}.</p>
   *
   * @param songEntry a SongEntry
   */
  default void notifyPlayed(@Nonnull SongEntry songEntry) {
    removeSuggestion(songEntry.getSong());
  }

  /**
   * <p>Removes the specified song from the {@link #getNextSuggestions(int) suggestions}.</p>
   *
   * @param song a song
   */
  void removeSuggestion(@Nonnull Song song);

  /**
   * <p>Indicates a user disliking the specified song.</p>
   *
   * The default implementation calls {@link #removeSuggestion(Song)}.
   *
   * @param song a song
   */
  default void dislike(@Nonnull Song song) {
    removeSuggestion(song);
  }

  /**
   * <p>Signals to the provider that it can start doing work.</p>
   *
   * <p>The Suggester should remain operational until {@link #close()} is called.</p>
   *
   * <p>The dependency Map is guaranteed to contain all required dependencies requested by {@link
   * #registerDependencies(DependencyReport)}. It will also contain all optional dependencies.</p>
   *
   * @param dependencies the Provider instances requested by {@link #registerDependencies(DependencyReport)}
   * @param initStateWriter a writer for initialization state messages
   * @throws InitializationException If any errors occur (for example due to invalid log-in
   * credentials)
   * @throws InterruptedException if the thread is interrupted while initializing
   */
  void initialize(@Nonnull InitStateWriter initStateWriter, @Nonnull DependencyMap<Provider> dependencies)
      throws InitializationException, InterruptedException;
}

File: src/main/kotlin/com/github/bjoernpetersen/musicbot/spi/plugin/Suggester.kt
package com.github.bjoernpetersen.musicbot.spi.plugin

import com.github.bjoernpetersen.musicbot.api.Song
import com.github.bjoernpetersen.musicbot.api.player.SongEntry

interface Suggester : Plugin, UserFacing, Active {

    /**
     * Suggest the next song to play.
     *
     * @return a song to play
     * @throws BrokenSuggesterException if the suggester can't suggest anything
     */
    @Throws(BrokenSuggesterException::class)
    fun suggestNext(): Song

    /**
     * Gets a list of next suggestions which will be returned by calling [suggestNext].
     *
     * The returned list must contain at least one element.
     *
     * @param maxLength the maximum length of the returned list
     * @return a list of next suggestions
     * @throws BrokenSuggesterException if the suggester can't suggest anything
     */
    @Throws(BrokenSuggesterException::class)
    fun getNextSuggestions(maxLength: Int): List<Song>

    /**
     * Notifies this Suggester that the specified song entry has been played.
     *
     * It is recommended that the song will be removed from the next suggestions when this method
     * is called.
     *
     * It is guaranteed that the song comes from a provider this suggester depends on.
     *
     * The default implementation calls [removeSuggestion].
     *
     * @param songEntry a SongEntry
     */
    fun notifyPlayed(songEntry: SongEntry) {
        removeSuggestion(songEntry.song)
    }

    /**
     *
     * Removes the specified song from the [suggestions][.getNextSuggestions].
     *
     * @param song a song
     */
    fun removeSuggestion(song: Song)

    /**
     * Indicates a user disliking the specified song.
     *
     * In contrast to a [removeSuggestion] call, this method indicates general dislike by the user
     * and that the song should be avoided in the future.
     *
     * The default implementation calls [removeSuggestion].
     *
     * @param song a song
     */
    fun dislike(song: Song) {
        removeSuggestion(song)
    }
}

/**
 * Indicates that a Suggester cannot make a suggestion request right now.
 *
 * This Exception does **not** imply that the suggester will be broken in the future.
 */
class BrokenSuggesterException : Exception {

    constructor() : super()
    constructor(message: String) : super(message)
    constructor(message: String, cause: Throwable) : super(message, cause)
    constructor(cause: Throwable) : super(cause)
}
